### DISCUSSION TITLE
Suppression Comments

### DISCUSSION THREAD

**Discussion by MichaReiser** [astral-sh/ruff#6338] [General]
This discussion proposes a design for suppression-comments in the formatter and documents the decision for the future. 

## Black 

Black supports `fmt: off` followed by an optional `fmt: on` own-line comments. Anything between does not get formatted: 

> Black reformats entire files in place. It doesn’t reformat lines that end with # fmt: skip or blocks that start with # fmt: off and end with # fmt: on. # fmt: on/off must be on the same level of indentation and in the same block, meaning no unindents beyond the initial indentation level between them. It also recognizes [YAPF](https://github.com/google/yapf)’s block comments to the same effect as a courtesy for straddling code. [[source](https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#code-style)]

```python
if outer:
 	if test:
  		# fmt:off
  		does    = not_get + formatted
  		# fmt:on
print("gets formatted")
```

The formatting gets automatically re-enabled if there's a dedent decreasing the indention level below the initial indentation of the `fmt: off` comment. The following is semantically identical to the above example:

```python
if outer:
 	if test:
  		# fmt:off
  		does    = not_get + formatted
print("gets formatted")
```

The indentation behavior doesn't apply to parenthesized expressions (any expression enclosed by a `()`, `[]`, or `{}` pair). Instead, Black re-enables the formatting when reaching the opening parentheses, meaning the following two examples are semantically equivalent:

```python
```python
a = (
	# fmt: off
	does +   not_get + formatted +
	# fmt: on
) + but + this + does

 
a = (
	# fmt: off
	does +   not_get + formatted +
) + but + this + does
```

Black further supports trailing `fmt: skip` comments that disable formatting for the preceding line. I haven't been able to figure out fully what the logic is when using inside of expressions. Here a few examples:

```python
# Doesn't format anything except that it adds two spaces before the comment
if (
        aaaa    < True
         + c): # fmt:skip
    bbb + c
    

# Formats everything, treads fmt:skip as a regular trailing comment?
if (
        aaaa    < True   # fmt:skip
         + c):
    bbb + c

# Formats content except the line annotated with `fmt:skip`
if (
    aaaa
    <
    (aaa        + b)  # fmt:skip
    + c
):
    bbb + c
```

## Proposal

Implementing Black's very flexible `fmt: off`..`fmt: on` suppression comments is challenging with our formatter architecture, and I also don't think it is necessary. I analyzed the usages of our ecosystem projects, and the vast majority only use statement-level suppression comments. These are easy to reason about and almost always sufficient. 

Suppression comments are also a sign that the formatter isn't good enough. I want to focus on improving formatting rather than investing in a very flexible and powerful suppression system. Suppression comments don't just mean that we're back to discussing how the suppressed code should be formatted, but the suppression also decreases readability overall. 

What's important from my perspective is that:

* You can intuitively tell which code gets formatted and how the suppressed code interacts with the enclosing formatted code (how the formatted file ultimately looks like)
* They support use cases like generated code 
* You can suppress any other code with reasonable effort (it may require assigning to temporary variables)

### Own line `fmt: off` and `fmt: on` comments

`fmt: off` and `fmt: on` comments allow to disable formatting for a range of statements. The statements must have the same or higher indentation level than the statement annotated with the `fmt: off` comment. The `fmt: on` is optional, but its best practice to use it. 

I propose to limit `fmt: off`... and `fmt: on` to statement level only. The following examples are all valid

```python
a = b

# fmt: off
not_formatted = 3
more
# fmt:on

if True:
  # fmt: off
  suite
  if and_nested:
    not_formatted

print("formatted")
```

However, the following examples are not valid (but are valid in black)

```python
if True:
	pass
# fmt: off
elif False:
	pass

@Test
# fmt: off
@Test2
def Abcd():
	pass
```

* The `if..elif` case doesn't fulfill the goal that suppression comments should be intuitive. Does it suppress the `elif` body only or all statements with the same indent as the `elif` case?
* The decorator case has the same problem. Does it disable the formatting only for the decorators, the decorators and the function signature, the decorators and the function, or the decorators, function, and any statements following after the function with the same indent? 

That's why I believe `fmt: skip` is useful

### Trailing `fmt: skip` or `fmt: off` comments

A trailing `fmt: skip` or `fmt: off` comment suppresses formatting for the preceding statement, case header, decorator, function definition, or class definition:

```python
if True:
	pass
elif False: # fmt: skip
	pass

@Test
@Test2 # fmt: off
def test(): ...

a = [1, 2, 3, 4, 5] # fmt: off

def test(a, b, c, d, e, f) -> int: # fmt: skip
	pass
```

There are two caveats where the use of `fmt: skip` is not intuitive:

```python
a; b; c # fmt: skip

def test(): pass # fmt: skip
```

* `a; b; c # fmt skip` suppresses the formatting for the statement `c`, and not the whole physical line as one might expect. This is because the AST has no sequence statement.
* `def test(): pass # fmt: skip` suppresses formatting for the whole function and not just the `pass` statement. 

I think this is a shortcoming that we have to accept.

## Unsupported Use cases

I went through all `fmt: off` and `fmt: skip` usages to identify the uses that either require changing the suppression comments or are entirely unsupported. 

### Between decorators

```python
huggingface:transformers/src/transformers/models/gpt2/modeling_gpt2.py
1516-
1517-        # Initialize weights and apply final processing
1518-        self.post_init()
1519-
1520-    @add_start_docstrings_to_model_forward(GPT2_INPUTS_DOCSTRING)
1521:    # fmt: off
1522-    @add_code_sample_docstrings(
1523-        checkpoint="brad1141/gpt2-finetuned-comp2",
1524-        output_type=TokenClassifierOutput,
1525-        config_class=_CONFIG_FOR_DOC,
1526-        expected_loss=0.25,
```

-> Use a trailing `fmt: skip`

### Inside Lists

```python
752-        vocab += [(piece.piece, piece.score) for piece in proto.pieces[3:]]
753-        vocab += [
754:            # fmt: off
755-            ('ace_Arab', 0.0), ('ace_Latn', 0.0), ('acm_Arab', 0.0), ('acq_Arab', 0.0), ('aeb_Arab', 0.0), ('afr_Latn', 0.0), ('ajp_Arab', 0.0), ('aka_Latn', 0.0), ('amh_Ethi', 0.0), ('apc_Arab', 0.0), ('arb_Arab', 0.0), ('ars_Arab', 0.0), ('ary_Arab', 0.0), ('arz_Arab', 0.0), ('asm_Beng', 0.0), ('ast_Latn', 0.0), ('awa_Deva', 0.0), ('ayr_Latn', 0.0), ('azb_Arab', 0.0), ('azj_Latn', 0.0), ('bak_Cyrl', 0.0), ('bam_Latn', 0.0), ('ban_Latn', 0.0), ('bel_Cyrl', 0.0), ('bem_Latn', 0.0), ...
756-            # fmt: on
757-        ]

jim22k:python-graphblas/graphblas/core/operator/unary.py
334-            UINT64,
335-        ]
336-        if _supports_complex:
337-            position_dtypes.extend([FC32, FC64])
338-        for names, *types in [
339:            # fmt: off
340-            (
341-                (
342-                    "erf", "erfc", "lgamma", "tgamma", "acos", "acosh", "asin", "asinh",
343-                    "atan", "atanh", "ceil", "cos", "cosh", "exp", "exp2", "expm1", "floor",
344-                    "log", "log10", "log1p", "log2", "round", "signum", "sin", "sinh", "sqrt",

jim22k:python-graphblas/graphblas/core/operator/semiring.py
409-        ]
410-        if _supports_complex:
411-            position_dtypes.extend([FC32, FC64])
412-            notbool_dtypes.extend([FC32, FC64])
413-        for lnames, rnames, *types in [
414:            # fmt: off
415-            (
416-                ("any", "max", "min", "plus", "times"),
417-                (
418-                    "firsti", "firsti1", "firstj", "firstj1",
419-                    "secondi", "secondi1", "secondj", "secondj1",

jim22k:python-graphblas/graphblas/core/operator/binary.py
743-            UINT64,
744-        ]
745-        if _supports_complex:
746-            position_dtypes.extend([FC32, FC64])
747-        name_types = [
748:            # fmt: off
749-            (
750-                ("atan2", "copysign", "fmod", "hypot", "ldexp", "remainder"),
751-                ((BOOL, INT8, INT16, UINT8, UINT16), FP32),
752-                ((INT32, INT64, UINT32, UINT64), FP64),
753-            ),

```

Either use `fmt: skip`, or assign the list to a temporary variable and suppress formatting for the assignment statement. 

### Inside assertions

```python
huggingface:transformers/tests/models/mbart50/test_tokenization_mbart50.py
84-        )
85-
86-        tokens = tokenizer.tokenize("I was born in 92000, and this is falsé.")
87-        self.assertListEqual(
88-            tokens,
89:            # fmt: off
90-            [SPIECE_UNDERLINE + "I", SPIECE_UNDERLINE + "was", SPIECE_UNDERLINE + "b", "or", "n", SPIECE_UNDERLINE + "in", SPIECE_UNDERLINE + "", "9", "2", "0", "0", "0", ",", SPIECE_UNDERLINE + "and", SPIECE_UNDERLINE + "this", SPIECE_UNDERLINE + "is", SPIECE_UNDERLINE + "f", "al", "s", "é", "."],
91-            # fmt: on
92-        )
93-        ids = tokenizer.convert_tokens_to_ids(tokens)
94-        self.assertListEqual(
--
100-        )

78-        self.assertListEqual(
79-            tokens,
80:            # fmt: off
81-            ['I', 'Ġwas', 'Ġborn', 'Ġin', 'Ġ9', '2000', ',', 'Ġand', 'Ġ', 'this', 'Ġis', 'Ġfals', 'Ã©', '.'],
82-            # fmt: on
83-        )
```

I've seen both `fmt: off` comments on the statement and expression level. Disallowing `fmt: off` on the expression level increases consistency. 

### If Conditions

```python
capyloon:gecko-b2g/third_party/python/pip/pip/_internal/commands/install.py
324-        target_temp_dir_path: Optional[str] = None
325-        if options.target_dir:
326-            options.ignore_installed = True
327-            options.target_dir = os.path.abspath(options.target_dir)
328-            if (
329:                # fmt: off
330-                os.path.exists(options.target_dir) and
331-                not os.path.isdir(options.target_dir)
332-                # fmt: on
333-            ):
334-                raise CommandError(
```

### Inside dictionaries

```python
huggingface:transformers/tests/pipelines/test_pipelines_token_classification.py
501-                "word": "##zo",
502-                "start": 2,
503-                "end": 4,
504-            },
505-            {
506:                # fmt: off
507-                "scores": np.array([0, 0, 0, 0.9986497163772583, 0]),
508-                # fmt: on
509-                "index": 7,
510-                "word": "UN",
511-                "is_subword": False,
--
557-                "word": "##zo",
558-                "start": 2,
559-                "end": 4,
560-            },
561-            {
562:                # fmt: off
563-                "scores": np.array([0, 0, 0, 0, 0, 0.9986497163772583, 0, 0, ]),
564-                # fmt: on
565-                "index": 7,
566-                "word": "UN",
567-                "is_subword": False,
```

This seems useful but won't be supported anymore. 

## Observation

The main use case for suppression comments is when it comes to formatting lists. We should improve list formatting by e.g. using a fill layout when all expressions in a list are simple expressions (numbers, simple strings). The fill layout tries to fit as many items as possible on a line rather than putting each item on its own line.

## Open Question

* Should we enforce/warn about missing  `fmt: on` comments?
* Is divering on the suppression comment an adoption blocker?

**Comment by charliermarsh** [comment]
FWIW, as I communicated in DMs, I'm supportive of this proposal.

Am I right that there's nothing in the proposed API that would _prevent_ us from expanding to support suppression-with-expressions in the future? (I'm ignoring the implementation cost and details -- I'm just confirming that there's nothing in the API that precludes us from adding that later if needed.)

**Reply by MichaReiser** [reply]
I don't see anything that would prevent us from supporting expression-level expressions in the future, beyond that, I don't know how to implement this in a sane way, and that I find expression level suppression comments unintuitive ;)

**Comment by zanieb** [comment]
Generally I am in support of this, it sounds great.

Are the following cases supported?

```python
[  # fmt: skip
   ...
]
```

and

```python
[...] # fmt: skip
```

Should `fmt: off` be recommended everywhere and just be treated differently if it's on its own line or trailing? Should we warn for use of `fmt: off` in places where `fmt: skip` should be used? Should we warn for unsupported usage of `fmt: off`?

> Should we enforce/warn about missing fmt: on comments?

Where would they be considered missing?

**Reply by MichaReiser** [reply]
> Are the following cases supported?
>  ```python
>  [  # fmt: skip
>       ...
>  ]
>  ```

No, this would not be supported because the comment is in the middle of a statement (expression even). Is this a pattern that you use frequently? 

> 
>  ```python
>  [...] # fmt: skip
>   ```

This is supported for as long the expression is at the end of the statement:

```python
# Supported, suppresses formatting of the whole statement
[ ... ] # fmt: skip
a = [...] # fmt: skip

# Unsupported
a = (
	[ ... ] # fmt: skip -> In the middle of an expression
)
```


> Should fmt: off be recommended everywhere and just be treated differently if it's on its own line or trailing? 

I'm leaning towards promoting `fmt: off` because:

* It increases the positions in which `fmt: off` comments are valid, making it less likely that they're used incorrectly. 
* You don't need to remember two suppression comment keywords. `fmt: off` is all you need.

I would still recommend supporting `fmt: skip` for backward compatibility (and even `yapf: disable`)

> Should we warn for use of fmt: off in places where fmt: skip should be used?

I wouldn't warn if we decide to support both (which is what I'm proposing)

> Should we warn for unsupported usage of fmt: off?

Yes, the same as we should warn about useless `fmt: on` comments. I'm not planning to implement this in the initial version, but we should warn eventually.

**Reply by MichaReiser** [reply]
> No, this would not be supported because the comment is in the middle of a statement (expression even). Is this a pattern that you use frequently?

To extend on why I propose not to allow additional places where `fmt: skip` is allowed, even though it would be technically straightforward to implement. The main motivation is that I thought about how we plan to teach this feature. Or in other words, how would the documentation for `fmt: skip` and `fmt: off..on` look like? 

Ideally, we come up with a short and precise definition for both. And that's where I believe we would struggle if we allow `fmt: skip` in some expression positions because we would need to enumerate them all (and people must remember them all). I already somewhat dislike that we need to allow them at the end of case headers and decorators because it requires us to use the jargon of `logical lines`: We allow them at the end of logical lines. 


Edit: 

Yet another reason is that I think it's unintuitive what `fmt: skip` suppresses in the following positions:

```python
a,    b =     [  # fmt: skip
     ...
]
```

Does the `# fmt: skip` suppress the formatting of everything up to the `[` or the list?

**Reply by zanieb** [reply]
> > Are the following cases supported?
> > ```python
> > [  # fmt: skip
> >      ...
> > ]
> > ```
> 
> No, this would not be supported because the comment is in the middle of a statement (expression even). Is this a pattern that you use frequently?

Not necessarily but that's how `# noqa` comment placement works, right?

> ```
> # Unsupported
> a = (
> 	[ ... ] # fmt: skip -> In the middle of an expression
> )
> ```

It might be nice to support this. It doesn't seem ambiguous as to what is being excluded here. Are there ambiguous cases?

> > Should fmt: off be recommended everywhere and just be treated differently if it's on its own line or trailing?
> 
> I'm leaning towards promoting `fmt: off` because...
> I would still recommend supporting `fmt: skip` for backward compatibility (and even `yapf: disable`)

+1 to that

**Reply by MichaReiser** [reply]
> Not necessarily but that's how # noqa comment placement works, right?

I'm not sure. Pragma comments are still a mystery to me. 

> It might be nice to support this. It doesn't seem ambiguous as to what is being excluded here. Are there ambiguous cases?

Not in this specific example, but it's probably rare to have a parenthesized list expression. It's more likely that you have something like this:

```python
a = {
	"key": [...] # fmt: skip
}

# Or

```python
a = (
	1 + 
	[ 2, 3,  4 ] # fmt: skip
)
```

Here I would expect that `fmt: skip` suppresses the list and the whole dictionary entry. The second one might only suppress the list formatting, but does that also mean that it enforces the list to remain on its own line or would 

```python
a = (
	1 + [2, 3, 4] # fmt: skip
)
```
be a valid formatting?

It's possible from a technical standpoint to support `fmt: skip` as leading or trailing node comments. This is what Rome and Prettier do. The main challenge I see with this approach is that *leading or trailing **node** comment* isn't something that's intuitive to users. How should I know where nodes start and end? I would need a tool telling me all valid suppression comment positions. It is probably true that this doesn't matter much in practice because most nodes start and end at the beginning of a line. The big exception to this are binary-like expressions and call chains

```python
(
	a
	+ b # fmt: skip
	+ c
)
```

This suppression happens to be in the middle of the `(a + b) + c` binary expression. Should it apply? If so, does it suppress the whole `a + b` expression, or only the `+ b` portion?

**Comment by MichaReiser** [comment]
To sum the discussion up to this point:

* There seems to be agreement on limiting `fmt: off..on` to the statement level only. I'll start working on this first
* It would be nice if `fmt: off/skip` comments would work on expression levels too. There are concerns about how to teach this, and it introduces the technical challenge that we need to get the expressions' parenthesized range.

**Comment by smackesey** [comment]
Would like to request that `# ruff: fmt:` works in addition to `# fmt:`-- this would match `# ruff: isort:` comments and make it more clear that they are ruff directives.

**Comment by peterjc** [comment]
I appreciate opinion will be divided over when and if arrays deserve manual layout, but trying out ruff 0.1.3, I see it does not obey examples like this (and related hand-tuned arrays formatted for element alignment):

```python
        self.assertTrue(
            np.array_equal(
                alignment.indices,
                np.array(
                    [
                        # fmt: off
                        [0,  1, 2,  3, 4, 5,  6, 7, -1, 8,  9, 10],
                        [0, -1, 1, -1, 2, 3, -1, 4,  5, 6, -1, -1],
                        # fmt: on
                    ]
                ),
            )
        )
```

Referring to the documentation above, I see the suggested alternatives for use with ``black`` or ``ruff format`` are:

* trailing ``# fmt: skip`` on each line of the array (does not scale well)
* trailing ``# fmt: skip`` on end of the statement (does not format rest of statement)
* not suggested above, but could also wrap entire statement in ``# fmt: off`` and ``# fmt: on`` 
* temporary variable, e.g.

```python
        # Up to user to decide how compact to make this array:
        expected = np.array([
                [0,  1, 2,  3, 4, 5,  6, 7, -1, 8,  9, 10],
                [0, -1, 1, -1, 2, 3, -1, 4,  5, 6, -1, -1],
        ])  # fmt: skip
        self.assertTrue(np.array_equal(alignment.indices, expected))

```

I'm guessing your preferred alternative for this use-case is the temporary variable?

(This example is actually a slight simplification as we also told flake8 to ignore the spaces; examples like this constitute the majority of the deviations from black in the repository concerned)

**Reply by MichaReiser** [reply]
Hy @peterjc 

Yes, I think you covered all options with the corresponding tradeoffs. I agree; these all feel less ergonomic than what Black and am sorry for that. 

We have multiple options from here:

1. Invest more time in supporting expression level `fmt: off` and `fmt: on`. An option is only to support them in specific situations to simplify the implementation and ensure that the formatted result is predictable (it hasn't always been the case for me when using Black). Although only supporting some expressions will be hard to teach, which is why we didn't go for it in the initial design.
2. Invest time in better supporting scientific expression, especially `np.array`, see https://github.com/astral-sh/ruff/discussions/7310#discussioncomment-7266904
3. Provide a suppression syntax that is easier to implement in Ruff, e.g. using `fmt: ignore` similar to Prettier which works on AST node level. 

Since we have limited time, I prefer to invest our time into the second option as soon as we finish our stabilization work on the formatter. The reasoning is that having to use `fmt: off` for `np.array` is already a workaround, and I would rather remove the need for the workaround than make the workaround better. If you have ideas on how to make `np.array` formatting better, please comment on https://github.com/astral-sh/ruff/discussions/7310#discussioncomment-7266904. A discussed option also is to automatically disable formatting of `np.array` for now until we can come up with something that works for the majority of users. 


Regarding your original question. I would probably suppress either the whole statement or extract the variable as you suggested.

**Comment by cjolowicz** [comment]
I didn't see table-driven tests with `pytest.mark.parametrize` mentioned, which was my main use case for `fmt: off`.

It turns out the trailing `fmt: skip` covers this perfectly:

```python
@pytest.mark.parametrize(
    ["which_result", "find_uv_bin_result", "expected"],
    [
        ("/usr/bin/uv", UV_IN_PIPX_VENV,   (True,  "/usr/bin/uv")),
        ("/usr/bin/uv", FileNotFoundError, (True,  "/usr/bin/uv")),
        (None,          UV_IN_PIPX_VENV,   (True,  UV_IN_PIPX_VENV)),
        (None,          FileNotFoundError, (False, "uv")),
    ],
)  # fmt: skip
def test_find_uv(monkeypatch, which_result, find_uv_bin_result, expected):
    ...
```
